bespoke-inc/deface
# add thresholds_by_sec & overlap_threshold for video_detect

#### What's this PR do?
This PR add customize `video_detect` function based on deface module.
It adds:
- thresholds_by_sec param: the threshold to identify a detected object as a face (and apply blurring). Now this param accept a dictionary of {time (second): float, threshold: float}
- Frame history of 5 previous detections.
- A consistency_threshold: the necessary number of frame in history that has overlap with new detection to consider a new detection is reliable.
- A grouping algorithm: to group all detected faces that have overlapping area into one detection (using weighted average centroids based on area and max width, height and score). This to avoid the case where detection are duplicated on the same face multiple time.

#### How should this be manually tested?
1. Unittest: almost all modified/ added function have their unittest. Run `python -m pytest tests/test_deface.py` to see if they are all passed

2. Installation & command line test: 
Install from this branch & run cli command to verify the new changes
```
pip install 'git+https://github.com/bespoke-inc/deface.git@phan/impv_video_detect'
OR
pip install -e .

deface examples/trimmed.mp4 --thresh 0.05 --thresholds_by_sec "{1: 0.99}"  --consistency_threshold 2
(after 1 sec, there should be no blurring) 
```

3. Calling in python code :
```
from deface.deface import video_detect
from deface.centerface import CenterFace
from pathlib import Path
import time

deface_path = "examples/trimmed.mp4"
threshold = 0.2

global_threshold = 0.05
local_thresholds = {1: 0.9}

video_detect(ipath=deface_path,
             opath="examples/trimmed_defaced.mp4", 
             default_threshold=global_threshold,
             keep_audio=True,
             centerface=CenterFace(in_shape=(640,480), backend="onnxrt"),
             replacewith='blur',
             mask_scale=1,
             ellipse=True,
             enable_preview=False,
             cam=False,
             nested=False,
             draw_scores=False,
             ffmpeg_config={"codec": "libx264"},
             thresholds_by_sec=local_thresholds,
             consistency_threshold=2
            )
```

#### Any background context you want to provide?
This PR is opened after discussion with Cyril [here](https://github.com/bespoke-inc/beproduced/pull/19)

#### Extra related tickets?
None

#### Open questions
None
